### DISCUSSION TITLE
using bitmap for tessellation

### DISCUSSION THREAD

**Discussion by majidkavousi** [neperfepx/neper#479] [Q&A] [answered]
Hi everyone

I plan to generate a microstructure from an Image using .tesr file/option. I have the image in .bmp and .png format (or any other common format) as below and want to create a cell for each colour. 
![untitled](https://user-images.githubusercontent.com/99205979/169876825-5a2a95dd-81bb-4dd6-a968-1800c4890c9c.png)
Does anyone have a similar experience that can help me? I really appreciate your help in advance.

**Comment by rquey** [comment, marked as answer]
The tesr format is defined [here](https://neper.info/doc/fileformat.html#raster-tessellation-file-tesr). You can get a sample tesr file using, e.g.,
```
neper -T -n 2 -dim 2 -for tesr
```

From the tesr file, you may reconstruct a tess using [-morpho "tesr:file(<file_name>)"](https://neper.info/doc/neper_t.html#cmdoption-morpho), but, if the final intent is to generate a mesh, such a 2D tesr can also be meshed directly:
```
neper -M <file_name>
```

**Reply by majidkavousi** [reply]
Thanks for your reply.
I have used the tesr format and been able to convert some images into meshes with quad elements. However, when I try to mesh with tri elements, I get the following error:

```
========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.4.1
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 1 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2022, and GNU GPL'd, by Romain Quey.
Info   : No initialization file found (`/home/majidkavousi/.neperrc').
Info   : ---------------------------------------------------------------
Info   : MODULE  -M loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] bigRVE.tesr -cl 0.004 -format msh,inp
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   :   - Reading arguments...
Info   : Loading input data...
Info   :   - Loading raster tessellation...
Info   :     [i] Parsing file `bigRVE.tesr'...
Info   :     [i] Parsing file `./bigRVE.raw'...
Info   :     [i] Parsed file `bigRVE.tesr'.
Info   :   - Reconstructing topology...
Info   :     > Reconstructing 1D mesh... 100%
Info   :     > Reconstructing 0D mesh... 100%
Error  : You have discovered a bug in Neper!  Please file an issue at
         https://github.com/rquey/neper/issues.  Thank you.
```

Do you have any idea about this issue?

**Reply by rquey** [reply]
Difficult to say for sure without the file, but you can try to run `-transform rmsat,grow` on your tesr; see https://github.com/neperfepx/neper/discussions/207#discussioncomment-698555.

**Reply by majidkavousi** [reply]
Hi Romain
Thanks for your kind answers. 
I've been working on the kind of problems I've mentioned in this discussion before and I've made good progress. However, at the moment I am getting errors while trying to mesh my microstructure with tri elements. There is no problem with quad elements. However, I need tri elements to be able to do some further modifications (like grouping, assigning different mesh sizes, mesh removal, etc.) I have attached a sample tesr file (with .txt format for being able to upload).
[RVE_361Gr.txt](https://github.com/neperfepx/neper/files/8975401/RVE_361Gr.txt)

```
$ neper -M RVE_361Gr.tesr -rcl 0.1

========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.4.1
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 1 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2022, and GNU GPL'd, by Romain Quey.
Info   : No initialization file found (`/home/majidkavousi/.neperrc').
Info   : ---------------------------------------------------------------
Info   : MODULE  -M loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] RVE_361Gr.tesr -rcl 0.1
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   :   - Reading arguments...
Info   : Loading input data...
Info   :   - Loading raster tessellation...
Info   :     [i] Parsing file `RVE_361Gr.tesr'...
Info   :     [i] Parsed file `RVE_361Gr.tesr'.
Info   :   - Reconstructing topology...
Info   :     > Reconstructing 1D mesh... 100%
Info   :     > Reconstructing 0D mesh... 100%
Error  : You have discovered a bug in Neper!  Please file an issue at
         https://github.com/rquey/neper/issues.  Thank you.
```

Another issue that I'm facing is that even when I try to mesh these raster tessellations with quad elements, I get Segmentation fault when I want to mesh with 2nd order elements.

```
$ neper -M RVE_361Gr.tesr -rcl 0.1 -elttype quad -order 2

========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.4.1
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 1 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2022, and GNU GPL'd, by Romain Quey.
Info   : No initialization file found (`/home/majidkavousi/.neperrc').
Info   : ---------------------------------------------------------------
Info   : MODULE  -M loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] RVE_361Gr.tesr -rcl 0.1 -elttype quad -order 2
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   :   - Reading arguments...
Info   : Loading input data...
Info   :   - Loading raster tessellation...
Info   :     [i] Parsing file `RVE_361Gr.tesr'...
Info   :     [i] Parsed file `RVE_361Gr.tesr'.
Info   : Meshing...
Info   :   - Preparing...

Info   :   - 2D meshing...
Info   :   - Switching mesh to order 2...
Segmentation fault (core dumped)
```

Do you have any recommendations on these errors?
Thanks

**Reply by rquey** [reply]
You [should have run](https://github.com/neperfepx/neper/discussions/479#discussioncomment-2873558) `-transform "rmsat,grow"`:

```
neper -T -loadtesr RVE_361Gr.tesr -tansform rmsat,grow -o d
neper -M d.tesr -order 2
neper -V d.tesr -print tesr
neper -V d.msh -dataelt2dedgerad 0.0002 -dataelt1drad 0.0006 -showelt1d all -print mesh
```
![image](https://user-images.githubusercontent.com/2659376/175527422-9bd34e8b-f66b-4367-8a1d-fbbf94be3671.png)
![image](https://user-images.githubusercontent.com/2659376/175527429-45e166af-cdeb-4876-9c83-caa55503b2bc.png)

**Reply by majidkavousi** [reply]
Thanks a lot. 
Now it works fine with tri elements.
